README.md
# SpeakReader

## What is SpeakReader
There is someone that attends my [church](https://www.ladoniabaptist.org) that is hearing impaired. We no longer have someone to sign for her. That was the inspiration for the development of this project.

There seems to be many speech-to-text applications out there. But they all seem to be designed for input and output on the same device. I wanted something that could output the speech-to-text transcription to many in realtime.

SpeakReader is a web server application that takes microphone or line-in input and transcribes it to text in realtime. It then sends the transcription text to listening web clients. Anyone with a smart device (smart phone, tablet, laptop, etc.) can connect to the web server and read what is being spoken.
 
SpeakReader can be used anywhere that spoken words need to be transcribed to text and shared to many such as churches, conferences, meetings, etc.
It is written in Python and can run on any platform that supports Python. As a very low cost solution, I am running SpeakReader in production at my church on a Raspberry Pi 3.

SpeakReader supports the following speech-to-text transcription services:
* [Google Speech-to-Text](https://cloud.google.com/speech-to-text/)
* [IBM Watson Speech-to-Text](https://www.ibm.com/watson/services/speech-to-text/)
* [Microsoft Azure Speech-to-Text](https://azure.microsoft.com/en-us/services/cognitive-services/speech-to-text/)

Not all transcription service providers are available on all platforms. For example, Microsoft Azure Speech Services is currently only available for Windows 32-bit and 64-bit, MacOS, and Linux 64-bit on non-ARM processors. 

If you find value in this project, please consider making a [donation via PayPal](https://paypal.me/jerryleenance). 80% of your donation will go to the [Ladonia Baptist Church](https://www.ladoniabaptist.org) building debt-retirement fund. 
You may also make a [donation](https://onrealm.org/LadoniaBaptist/Give/EAVLVGBZJN) directly to the church building fund through their online-giving system.

## Requirements
SpeakReader requires Python 3.6 or higher.


## Installation
[Installation instructions](https://github.com/zSeriesGuy/SpeakReader/wiki) can be found in the wiki.
 

## Usage

The default port for SpeakReader after installation is 8880. This can be changed in the configuration on the settings page or by editing the config.ini file.

Connect a microphone or line-in to the server that you are running SpeakReader on and start SpeakReader.
The recommendation is to have a good clean input feed. For my production server, I use a line feed directly off of the church sound board. The cleaner the input feed, the more accurate the transcription results will be.

The transcription services do not have to be "trained" to the speaker. It gets the majority of the words correct. But due to differences in speaking style, clarity, accents, the transcription services may not get every word correct.

Profanity filters have been enabled by default in the transcription services to mask any intentional or unintentional profanity. This can be changed on the settings page as well as editing a custom list of censored words. 


#### User Interface
Access the web server. Be sure to include the port if you are not using the standard 80 or 443 (*`http://speakreaderURL:8880`*).

The home page on the SpeakReader web server is the transcription text.
The web page automatically scrolls to bottom. To pause auto scroll, just scroll back in the transcription text. A bottom button will appear in the upper right that, when clicked, will resume auto scrolling at the bottom position in the transcript.

To adjust the font, tap or click in the transcription window. A tab will slide up from the bottom allowing font size and color theme adjustment.

#### Management Console
To manage SpeakReader, access the web server's **`/manage`** page.

There are four tabs on the management page: **Status**, **Settings**, **Transcripts**, and **Logs**.

##### Status Page
The Status page:
* Shows the current status of the transcription engine.
* Shows the current input device.
* Allows for starting and stopping the transcription engine.
* Allows for restarting and shutting down the SpeakReader server.
* View list of connected listeners
* Monitor the active transcript or console log.

##### Settings Page
The Settings page allows you to make and save changes to the configuration which are stored in the config.ini file.

Settings include:
* Select the microphone or line-in input device.
* Create a userid/password for restricting access to the management console.
* Upload your GoogleAPI credentials file.
* Whether or not to start the transcribe engine on server start up.
* Whether or not to launch a web browser to the server on server start up.
* Set the folder where to store transcripts.
* Set the port the server will listen on.
* Enable HTTPS for secure communications.
* Enable/disable censorship and add additional words to the censored words list.

##### Transcripts Page
The Transcripts page shows a list of all transcript files in the Transcripts folder. From the list, you can view, download, or delete the transcript.

##### Logs Page
The Logs page shows a list of all log files in the Logs folder. From the list, you can view, download, or delete the log.


CHANGELOG.md
# Changelog

## v1.2.7 (2021-05-20)
* Add PyAudio wheels for Python 3.9.
* Upgrade packages, specifically RSA to address security vulnerability.
* Changes to support google-cloud-speech V2.
* Add API service name in use to the TranscribeEngine startup message.

## v1.2.6 (2020-05-21)
* Upgrade packages, specifically httplib2 to address security vulnerability.
* Change how Updates Available is presented in Settings when there are commit updates.
* Add pip-update method to version management to update the python installation tools during update.

## v1.2.5 (2019-12-10)
* Upgrade packages, specifically grpcio, to fix an issue on Raspbian.
* Modify pip_sync to output messages from process in realtime.
* Modify loop to handle Google Cloud time limits on content.
* Set GIT remote and branch based on what's installed.

## v1.2.4 (2019-12-09)
* Fix favicon handling.
* Run pip install requirements before pip-sync.
* Update installation instructions and init scripts

## v1.2.3 (2019-10-21)
* Add support to ensure only one instance of SpeakReader is running.
* Add exception handler for Google OutOfRange exception. Streaming limit is 5 minutes.
* Run file cleanup if days changes in management console settings.
* Upgrade packages.

## v1.2.2 (2019-10-19)
* Add settings and process for deleting logs, transcripts, and recordings after a specified number of days.

## v1.2.1 (2019-10-19)
* Microsoft Azure Speech-to-Text Cloud service is not available on some platforms. Added support for detecting failure to load service.

## v1.2.0 (2019-10-15)
* Add support for Microsoft Azure Speech-to-Text Cloud service.
* Remove Nightly as a selectable git branch.

## v1.1.0 (2019-10-15)
**Changes:**
* Modifications to support different transcription services.
* Update packages.

**New:**
* Add support for IBM Watson Speech-to-Text Cloud service.

> **NOTE:** Due to the modifications to support multiple transcription services, if you are upgrading from a previous release, you will need to re-upload your Google API JSON file or edit the config.ini file and change credentials_file to google_credentials_file.

## v1.0.5 (2019-10-15)
* Run pip-sync on git branch change.

## v1.0.4 (2019-10-08)
* Increase max size of transcript font setting.
* Correct issues using IOS Safari browser.

## v1.0.3 (2019-09-15)
* Added db sound meter.

## v1.0.2 (2019-09-01)
* Resample input device stream to 16kHz.
* Added saving the sound recording to a wav file.

## v1.0.1 (2019-08-31)
* Change to use input device's default sample rate.

## v1.0.0 (2019-08-15)
* Initial Release


start.py
#!/bin/sh
''''which python3   >/dev/null 2>&1 && exec python    "$0" "$@" # '''
''''exec echo "Error: Python3 not found!" # '''

# **************************************************************************************
# * This file is part of SpeakReader.
# *
# *  SpeakReader is free software: you can redistribute it and/or modify
# *  it under the terms of the GNU General Public License V3 as published by
# *  the Free Software Foundation.
# *
# *  SpeakReader is distributed in the hope that it will be useful,
# *  but WITHOUT ANY WARRANTY; without even the implied warranty of
# *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# *  GNU General Public License for more details.
# *
# *  You should have received a copy of the GNU General Public License
# *  along with SpeakReader.  If not, see <http://www.gnu.org/licenses/gpl-3.0.html>.
# **************************************************************************************

#
# This file is the main startup module for SpeakReader.
#

import os
import sys
import subprocess
from platform import system, release, version, machine, processor

import argparse
import datetime
import locale
import pytz
import signal
import time
import tzlocal
import threading
import psutil

import speakreader
from speakreader import logger, config, SpeakReader

# Ensure lib added to path, before any other imports
sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), 'lib'))


def main():
    """
    SpeakReader application entry point. Parses arguments, setups encoding and
    initializes the application.
    """

    DAEMON = False
    NOFORK = False

    QUIET = False
    VERBOSE = False
    NOLAUNCH = False
    HTTP_PORT = None

    PLATFORM = system()
    PLATFORM_RELEASE = release()
    PLATFORM_VERSION = version()
    PLATFORM_LINUX_DISTRO = None

    PLATFORM_PROCESSOR = processor()
    PLATFORM_MACHINE = machine()
    PLATFORM_IS_64BITS = sys.maxsize > 2**32
    SYS_PLATFORM = sys.platform

    # Fixed paths to application
    if hasattr(sys, 'frozen'):
        FULL_PATH = os.path.abspath(sys.executable)
    else:
        FULL_PATH = os.path.abspath(__file__)

    PROG_DIR = os.path.dirname(FULL_PATH)

    # Ensure only one instance of SpeakReader running.
    PIDFILE = os.path.join(PROG_DIR, 'pidfile')
    myPid = os.getpid()
    if os.path.exists(PIDFILE) and os.path.isfile(PIDFILE):
        for p in psutil.process_iter():
            if 'python' in p.name() and p.pid != myPid:
                for f in p.open_files():
                    if f.path == PIDFILE:
                        logger.error("SpeakReader is already Running. Exiting.")
                        sys.exit(0)

    myPidFile = open(PIDFILE, 'w+')
    myPidFile.write(str(myPid))
    myPidFile.flush()

    try:
        locale.setlocale(locale.LC_ALL, "")
    except (locale.Error, IOError):
        pass

    try:
        SYS_TIMEZONE = str(tzlocal.get_localzone())
        SYS_UTC_OFFSET = datetime.datetime.now(pytz.timezone(SYS_TIMEZONE)).strftime('%z')
    except (pytz.UnknownTimeZoneError, LookupError, ValueError) as e:
        logger.error("Could not determine system timezone: %s" % e)
        SYS_TIMEZONE = 'Unknown'
        SYS_UTC_OFFSET = '+0000'

    # Parse any passed startup arguments
    ARGS = sys.argv[1:]
    # Set up and gather command line arguments
    parser = argparse.ArgumentParser(description='A Python based monitoring and tracking tool for Plex Media Server.')

    parser.add_argument(
        '-v', '--verbose', action='store_true', help='Increase console logging verbosity')
    parser.add_argument(
        '-q', '--quiet', action='store_true', help='Turn off console logging')
    parser.add_argument(
        '-d', '--daemon', action='store_true', help='Run as a daemon')
    parser.add_argument(
        '-p', '--port', type=int, help='Force SpeakReader to run on a specified port')
    parser.add_argument(
        '--datadir', help='Specify a directory where to store your data files')
    parser.add_argument(
        '--config', help='Specify a config file to use')
    parser.add_argument(
        '--nolaunch', action='store_true', help='Prevent browser from launching on startup')
    parser.add_argument(
        '--nofork', action='store_true', help='Start SpeakReader as a service, do not fork when restarting')

    args = parser.parse_args()

    # Force the http port if necessary
    if args.port:
        HTTP_PORT = args.port
        logger.info('Using forced web server port: %i', HTTP_PORT)

    # Don't launch the browser
    if args.nolaunch:
        NOLAUNCH = True

    if args.verbose:
        VERBOSE = True
    if args.quiet:
        QUIET = True

    if args.daemon:
        if SYS_PLATFORM == 'win32':
            sys.stderr.write("Daemonizing not supported under Windows, starting normally\n")
        else:
            DAEMON = True
            QUIET = True

    if args.nofork:
        NOFORK = True
        logger.info("SpeakReader is running as a service, it will not fork when restarted.")

    # Determine which data directory and config file to use
    if args.datadir:
        DATA_DIR = args.datadir
    else:
        DATA_DIR = os.path.join(PROG_DIR, 'data')

    if args.config:
        CONFIG_FILE = args.config
    else:
        CONFIG_FILE = os.path.join(DATA_DIR, config.FILENAME)

    # Try to create the DATA_DIR if it doesn't exist
    if not os.path.exists(DATA_DIR):
        try:
            os.makedirs(DATA_DIR)
        except OSError:
            raise SystemExit('Could not create data directory: ' + DATA_DIR + '. Exiting....')

    # Make sure the DATA_DIR is writeable
    if not os.access(DATA_DIR, os.W_OK):
        raise SystemExit('Cannot write to the data directory: ' + DATA_DIR + '. Exiting...')

    # Do an initial setup of the logging.
    logger.initLogger(console=not QUIET, log_dir=False, verbose=VERBOSE)

    # Initialize the configuration from the config file
    CONFIG = config.Config(CONFIG_FILE)
    assert CONFIG is not None

    if CONFIG.SERVER_ENVIRONMENT.lower() != 'production':
        VERBOSE = True

    CONFIG.LOG_DIR, log_writable = check_folder_writable(
        CONFIG.LOG_DIR, os.path.join(DATA_DIR, 'logs'), 'logs')
    if not log_writable and not QUIET:
        sys.stderr.write("Unable to create the log directory. Logging to screen only.\n")

    # Start the logger, disable console if needed
    logger.initLogger(console=not QUIET, log_dir=CONFIG.LOG_DIR if log_writable else None, verbose=VERBOSE)

    logger.info("Initializing {} {}".format(speakreader.PRODUCT, speakreader.VERSION_RELEASE))
    logger.info("{} {} ({}{})".format(
        PLATFORM, PLATFORM_RELEASE, PLATFORM_VERSION,
        ' - {}'.format(PLATFORM_LINUX_DISTRO) if PLATFORM_LINUX_DISTRO else ''
    ))
    logger.info("{} ({} {})".format(
        PLATFORM_PROCESSOR, PLATFORM_MACHINE,
        '{}'.format('64-BIT') if PLATFORM_IS_64BITS else '32-BIT'
    ))
    logger.info("Python {}".format(
        sys.version
    ))
    logger.info("{} (UTC{})".format(
        SYS_TIMEZONE, SYS_UTC_OFFSET
    ))
    logger.info("Program Dir: {}".format(
        PROG_DIR
    ))
    logger.info("Config File: {}".format(
        CONFIG_FILE
    ))

    CONFIG.TRANSCRIPTS_FOLDER, _ = check_folder_writable(
        CONFIG.TRANSCRIPTS_FOLDER, os.path.join(DATA_DIR, 'transcripts'), 'transcripts')

    CONFIG.RECORDINGS_FOLDER, _ = check_folder_writable(
        CONFIG.RECORDINGS_FOLDER, os.path.join(DATA_DIR, 'recordings'), 'recordings')

    if DAEMON:
        daemonize(myPidFile)

    # Store the original umask
    UMASK = os.umask(0)
    os.umask(UMASK)

    initOptions = {
        'config': CONFIG,
        'http_port': HTTP_PORT,
        'nolaunch': NOLAUNCH,
        'prog_dir': PROG_DIR,
        'data_dir': DATA_DIR,
    }

    # Read config and start logging
    global SR
    SR = SpeakReader(initOptions)

    # Wait endlessly for a signal to happen
    restart = False
    checkout = False
    update = False
    while True:
        if not SR.SIGNAL:
            try:
                time.sleep(1)
            except KeyboardInterrupt:
                SR.SIGNAL = 'shutdown'
        else:
            logger.info('Received signal: %s', SR.SIGNAL)

            if SR.SIGNAL == 'shutdown':
                break
            elif SR.SIGNAL == 'restart':
                restart = True
                break
            elif SR.SIGNAL == 'update':
                restart = True
                update = True
                break
            elif SR.SIGNAL == 'checkout':
                restart = True
                checkout = True
                break
            else:
                SR.SIGNAL = None

    SR.shutdown(restart=restart, update=update, checkout=checkout)

    myPidFile.close()
    os.remove(PIDFILE)

    if restart:
        logger.info("SpeakReader is restarting...")

        exe = sys.executable
        args = [exe, FULL_PATH]
        args += ARGS
        if '--nolaunch' not in args:
            args += ['--nolaunch']

        # Separate out logger so we can shutdown logger after
        if NOFORK:
            logger.info('Running as service, not forking. Exiting...')
        elif os.name == 'nt':
            logger.info('Restarting SpeakReader with %s', args)
        else:
            logger.info('Restarting SpeakReader with %s', args)

        logger.shutdown()

        if NOFORK:
            pass
        elif os.name == 'nt':
            subprocess.Popen(args, cwd=os.getcwd())
        else:
            os.execv(exe, args)

    else:
        logger.info("SpeakReader Terminated")
        logger.shutdown()

    sys.exit(0)


def daemonize(myPidFile):

    if threading.activeCount() != 1:
        logger.warn(
            "There are %r active threads. Daemonizing may cause"
            " strange behavior.",
            threading.enumerate())

    sys.stdout.flush()
    sys.stderr.flush()

    # Do first fork
    try:
        pid = os.fork()  # @UndefinedVariable - only available in UNIX
        if pid != 0:
            sys.exit(0)
    except OSError as e:
        raise RuntimeError("1st fork failed: %s [%d]", e.strerror, e.errno)

    os.setsid()

    # Make sure I can read my own files and shut out others
    prev = os.umask(0)  # @UndefinedVariable - only available in UNIX
    os.umask(prev and int('077', 8))

    # Make the child a session-leader by detaching from the terminal
    try:
        pid = os.fork()  # @UndefinedVariable - only available in UNIX
        if pid != 0:
            sys.exit(0)
    except OSError as e:
        raise RuntimeError("2nd fork failed: %s [%d]", e.strerror, e.errno)

    dev_null = open('/dev/null', 'r')
    os.dup2(dev_null.fileno(), sys.stdin.fileno())

    si = open('/dev/null', "r")
    so = open('/dev/null', "a+")
    se = open('/dev/null', "a+")

    os.dup2(si.fileno(), sys.stdin.fileno())
    os.dup2(so.fileno(), sys.stdout.fileno())
    os.dup2(se.fileno(), sys.stderr.fileno())

    pid = os.getpid()
    myPidFile.seek(0)
    myPidFile.write(str(pid))
    myPidFile.flush()
    logger.info("Daemonized to PID: %d", pid)


def check_folder_writable(folder, fallback, name):
    if not folder:
        folder = fallback

    if not os.path.exists(folder):
        try:
            os.makedirs(folder)
        except OSError as e:
            logger.error("Could not create %s dir '%s': %s" % (name, folder, e))
            if folder != fallback:
                logger.warn("Falling back to %s dir '%s'" % (name, fallback))
                return check_folder_writable(None, fallback, name)
            else:
                return folder, None

    if not os.access(folder, os.W_OK):
        logger.error("Cannot write to %s dir '%s'" % (name, folder))
        if folder != fallback:
            logger.warn("Falling back to %s dir '%s'" % (name, fallback))
            return check_folder_writable(None, fallback, name)
        else:
            return folder, False

    return folder, True


def sig_handler(signum=None, frame=None):
    global SR
    if signum is not None:
        logger.info("sig_handler: Signal %i caught, Shutting Down SpeakReader...", signum)
        SR.SIGNAL = 'shutdown'


# Register signals, such as CTRL + C
signal.signal(signal.SIGINT, sig_handler)
signal.signal(signal.SIGTERM, sig_handler)


# Call main()
if __name__ == "__main__":
    main()
